Add vtk9 support (phase4)

biagas opened the pull request:
### Description

This PR focuses on the change to dataset cell access, requiring 'const vtkIdType*'. (GetCellPoints, GetNextCell, InsertNextCell, GetCell).
GetCell  changed to  GetCellAtId
vtkCellLinks::Link::nCells changed from unsigned short to vtkIdType.

### Type of change

<!-- Please check one of the boxes below -->

* [ ] Bug fix~~
* [X] New feature~~
* [ ] Documentation update~~
* [ ] Other~~ <!-- please explain with a note below -->

### How Has This Been Tested?

I compiled on Windows and Linux and ran the serial version of the test suite on pascal with success.

### Checklist:

<!-- For items in this checklist that do not apply, simply insert two tilde chars, `~~`, just ahead of the left bracket char, `[` at the beginning of a line. Each line ends with two tilde chars to make doing such ~~strikeouts~~ easy. -->

- [X] I have followed the [style guidelines][1] of this project.~~
- [X] I have performed a self-review of my own code.~~
~~- [ ] I have commented my code where applicable.~~
~~- [ ] I have updated the release notes.~~
~~- [ ] I have made corresponding changes to the documentation.~~
~~- [ ] I have added debugging support to my changes.~~
~~- [ ] I have added tests that prove my fix is effective or that my feature works.~~
- [X] I have confirmed new and existing unit tests pass locally with my changes.~~
~~- [ ] I have added new baselines for any new tests to the repo.~~
~~- [ ] I have NOT made any changes to [*protocol* or *public interfaces*][3] in an RC branch.~~
- [X] I have assigned reviewers (see [VisIt's PR procedures][2] for more information).~~

[1]: https://visit-sphinx-github-user-manual.readthedocs.io/en/develop/dev_manual/StyleGuide.html
[2]: https://visit-sphinx-github-user-manual.readthedocs.io/en/develop/dev_manual/pr_create.html#reviewers
[3]: https://visit-sphinx-github-user-manual.readthedocs.io/en/develop/dev_manual/RCDevelopment.html#communication-protocols-and-public-apis

biagas committed:
VTK 9 change to use 'const vtkIdType *` for some cell array accessors.

GetCellPoints,
GetNextCell

biagas committed:
Merge branch 'develop' into task/biagas/add_vtk9_support_phase4

biagas committed:
Forgot vtkVisItTubeFilter header

biagas committed:
Add `const vtkIdType *` fix for avtTubeFilter.C

biagas committed:
Merge branch 'develop' into task/biagas/add_vtk9_support_phase4

biagas committed:
Merge branch 'develop' into task/biagas/add_vtk9_support_phase4

markcmiller86 commented on src/avt/Filters/avtLineToPolylineFilter.C:
```diff
@@ -157,28 +159,33 @@ avtLineToPolylineFilter::ExecuteData(avtDataRepresentation *inDR)
     std::set<edge> freeEdges;
     vtkIdType toCellId = input->GetVerts()->GetNumberOfCells();
-    for(vtkIdType cellid = 0; input->GetLines()->GetNextCell(n, pts); ++cellid)
+    for(vtkIdType cellid = 0; input->GetLines()->GetNextCell(n, cellPts); ++cellid)
```
I was completely perplexed by the old code's behavior w.r.t. to `pts` and by association the new code w.r.t. `cellPts`. In the old code, I kept looking for some place `pts` is either allocated or freed (or was thinking we might have a leak). But, it looks like what `GetNextCell` does is just update a pointer to some internal object data and it should not ever be modified/freed. Thats probably expedient from a performance perspective but threw me for a loop in reading this block of code.

biagas replied:
In this case I changed 'pts' to 'cellPts' because 'pts' is used again later in a different context.

Looking again at the documentation, this probably should have been changed to use vtkCellArrayIterator for VTK9, a better way of traversing the vtkCellArray (which is what input->GetLines() returns).  VTK 9 documentation states the method isn't thread safe.  I hadn't noticed that before.

Putting this PR in draft so I can reconsider the changes.

biagas replied:
Can't figure out how to put it in draft mode :(

markcmiller86 replied:
![Screen Shot 2022-08-12 at 2 05 51 PM](https://user-images.githubusercontent.com/5720676/184444945-d8984f51-b545-460c-a6b2-b7ccdc7602d7.png)

markcmiller86 commented on src/avt/Plotter/avtTransparencyActor.C:
```diff
@@ -35,6 +37,7 @@
 #include <vtkParallelImageSpaceRedistributor.h>
 #include <vtkTransform.h>
 #include <vtkTransformFilter.h>
+#include <vtkUnsignedCharArray.h>
```
New header here but no diffs below indicate it is needed. Still needed?

biagas replied:
vtkUnsignedCharArray is used in the file and the #include is necessary.

I think it used to be pulled in from some other headers.

markcmiller86 replied:
Ah, ok. So we were *missing* including that header before and just getting lucky. Ok? Probably same for all the other instances I pointed out too. I didn't think to go look for that.

markcmiller86 commented on src/avt/Queries/Misc/avtDirectDatabaseQOTFilter.C:
```diff
@@ -577,11 +580,16 @@ avtDirectDatabaseQOTFilter::VerifyAndRefineArrayTimesteps(
     // that may have occured during coordinate retrieval.
     //
-    for (int i, ts = 0; i < numCells; ++i, ts += stride)
+    for (int i = 0, ts = 0; i < numCells; ++i, ts += stride)
```
Seems like orig. code was an uninitialized loop index variable. Did this ever work before you changes?

biagas replied:
Not sure, I know we have tests.  I think with linux/gcc we get lucky and uninitialized int is 'unofficially' inited to '0'.

markcmiller86 commented on src/databases/SimV2/avtSimV2Writer.C:
```diff
@@ -13,6 +15,7 @@
 #include <vtkPolyData.h>
 #include <vtkRectilinearGrid.h>
 #include <vtkStructuredGrid.h>
+#include <vtkUnsignedCharArray.h>
```
Same question here. Header was added but no diffs to code below indicating it was needed.

markcmiller86 commented on src/operators/Isovolume/avtIsovolumeFilter.C:
```diff
@@ -8,22 +8,26 @@
 
-#include <vtkVisItClipper.h>
-#include <vtkDataSet.h>
+#include <vtkCellData.h>
```
Another instance of new headers but no diffs indicating anything new required them. I must be misunderstanding.

markcmiller86 commented on src/visit_vtk/full/vtkAxisDepthSort.C:
```diff
@@ -152,7 +153,11 @@ vtkAxisDepthSort::RequestData(
+#if LIB_VERSION_LE(VTK, 8,1,0)
     vtkIdType    *cellPts    = 0;
+#else
+    const vtkIdType *cellPts = nullptr;
```
Can you use `nullptr` above to make it consistent?

markcmiller86 commented on src/visit_vtk/full/vtkAxisDepthSort.h:
```diff
@@ -44,15 +46,15 @@ class VISIT_VTK_API vtkAxisDepthSort : public vtkPolyDataAlgorithm
-                                  vtkAxisDepthSort(const vtkAxisDepthSort &);
-                                  void operator=(const vtkAxisDepthSort &);
+                                  vtkAxisDepthSort(const vtkAxisDepthSort &) = delete;
+                                  void operator=(const vtkAxisDepthSort &) = delete;
```
Is `private` still necessary if using `= delete;`

biagas replied:
Unclear to me. The change to this file was to be consistent with other files that had previously been using 'VTK_DELETE' and needed to be changed to 'delete'.

markcmiller86 replied:
Ah, ok. I think I was misunderstanding then. Does `= delete` actually remove them from the vtable? If so, that is indeed something different than making the private (they're in the vtable but just not callable outside). Honestly, I am just guessing here as this is the first time I've ever seen this construct.

biagas replied:
Introduced in C++11, to avoid implicit declaration of a copy constructor or assignment operator.
https://en.cppreference.com/w/cpp/language/copy_constructor

markcmiller86 commented on src/visit_vtk/full/vtkVisItTubeFilter.C:
```diff
@@ -104,7 +104,12 @@ int vtkVisItTubeFilter::RequestData(
+    vtkIdType npts=0;
+#if LIB_VERSION_LE(VTK, 8,1,0)
+    vtkIdType *pts=NULL;
+#else
```
`NULL` vs `nullptr` here. Does it matter? Can we make consistent?

biagas replied:
Should I changes all instances of 'NULL' in the file to 'nullptr' ?

markcmiller86 replied:
Though that might not be a bad idea, I am also not suggesting that either. But these two are really closely related and I thought it feels off to have they use different values for initialization.

markcmiller86 commented on src/visit_vtk/full/vtkVisItTubeFilter.h:
```diff
@@ -233,8 +247,8 @@ class VISIT_VTK_API vtkVisItTubeFilter : public vtkPolyDataAlgorithm
-  vtkVisItTubeFilter(const vtkVisItTubeFilter&);  // Not implemented.
-  void operator=(const vtkVisItTubeFilter&);  // Not implemented.
+  vtkVisItTubeFilter(const vtkVisItTubeFilter&) = delete;
+  void operator=(const vtkVisItTubeFilter&) = delete;
```
Same question as above...is `private:` scope still needed when using `= delete`. 

BTW...this seems like a new style of excluding such methods and maybe worth a quick announcement or something. I would have never known this is the new way of doing business.

markcmiller86 commented on src/visit_vtk/lightweight/vtkPolyDataRelevantPointsFilter.C:
```diff
@@ -203,7 +203,11 @@ int vtkPolyDataRelevantPointsFilter::RequestData(
   vtkIdType *pts = new vtkIdType[nIdStoreSize];
+#if LIB_VERSION_LE(VTK, 8,1,0)
   vtkIdType *oldPts = NULL;
+#else
```
Another instance of `NULL` vs `nullptr`.

markcmiller86 reviewed (commented):
Looks good. I did point out a few things I wasn't sure about.

biagas committed:
Use vtkCellArrayIterator for cell traversal with VTK 9.

Use GetCellAtId for VTK9 instead of GetCell, as GetCell is documented to be very slow with VTK 9.
Use nullptr a bit more consistently.

biagas committed:
Merge branch 'develop' into task/biagas/add_vtk9_support_phase4

biagas committed:
Merge branch 'develop' into task/biagas/add_vtk9_support_phase4
